tawazi/_dag/helpers.py

```diff
@@ -0,0 +1,298 @@
+        xns_dict (Dict[Identifier, ExecNode]): The ExecNodes whose Futures are submitted.
+
+    Returns:
+        Optional[float]: The next timeout to use in wait.
+    """
+    # start next_timeout as infinity.
+    # 1st timeout encountered will reduce it to a finite number.
+    next_timeout = float("inf")
+    for id_, fut in futures.items():
+        # if future is not done, then it should be running (This is not guaranteed in ThreadPoolExecutor)
+        if not fut.done():
+            launch_time = futures_launch_time[id_]
+            timeout = xns_dict[id_].timeout
+            logger.debug(
+                f"next_timeout: {next_timeout}, launch_time: {launch_time}, timeout: {timeout}"
+            )
+
+            if timeout is not None:
+                next_timeout = min(next_timeout, launch_time + timeout)
+
+    if next_timeout == float("inf"):
+        return None
+    return next_timeout - monotonic()
+
+
+def raise_timeout_error_conditionally(
+    done_: Set["Future[Any]"], running: Set["Future[Any]"], next_timeout: Optional[float]
+) -> None:
+    # done_ doesn't contain any future. => timeout reached
+    if not done_:
+        raise TawaziTimeoutError(
+            f"Timeout reached while waiting for {running} to finish. Waited a total of {next_timeout} seconds"
+        )
+
+
+################
+# The scheduler!
+################
+def execute(
+    *,
+    node_dict: Dict[Identifier, ExecNode],
+    max_concurrency: int,
+    behavior: ErrorStrategy,
+    graph: DiGraphEx,
+    modified_node_dict: Optional[Dict[str, ExecNode]] = None,
+    call_id: str = "",
+) -> Dict[Identifier, Any]:
+    """Thread safe execution of the DAG.
+
+    (Except for the setup nodes! Please run DAG.setup() in a single thread because its results will be cached).
+
+    Args:
+        node_dict: dictionary identifying ExecNodes.
+        max_concurrency: maximum number of threads to be used for the execution.
+        behavior: the behavior to be used in case of error.
+        graph: the graph ids to be executed
+        modified_node_dict: A dictionary of the ExecNodes that have been modified by setting the input parameters of the DAG.
+        call_id (str): A unique identifier for the execution.
+            It can be used to distinguish the id of the call inside the thread.
+            It might be useful to debug and to exchange information between the main thread and the sub-threads (per-node threads)
+
+    Returns:
+        node_dict: dictionary with keys the name of the function and value the result after the execution
+    """
+    # 0.1 deepcopy the node_dict in order to modify the results inside every node and make the dag reusable
+    #     modified_node_dict are used to modify the values inside the ExecNode corresponding
+    #     to the input arguments provided to the whole DAG (ArgExecNode)
+    xns_dict = modified_node_dict or copy_non_setup_xns(node_dict)
```

Comment:
not a huge fan of `or` between two variables that are not explicit booleans, would you mind adding `if xxx is not None` at the end of the line ?

Reply:
I will fix this in a separate PR

Reply:
okok